File: src/main/java/com/github/dylmeadows/common/sound/SoundPlayer.java
package com.github.dylmeadows.common.sound;

import com.github.dylmeadows.eontimer.model.resources.SoundResource;
import com.github.dylmeadows.eontimer.util.extension.OptionalExtensions;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.scene.media.Media;
import javafx.scene.media.MediaPlayer;
import javafx.util.Duration;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

import java.util.Optional;

/**
 * Helper class for loading and playing audio clips.
 */
@Slf4j
public class SoundPlayer {

    private MediaPlayer mediaPlayer;

    private ObjectProperty<SoundResource> sound = new SimpleObjectProperty<>();

    public void play() {
        OptionalExtensions.ifPresentOrElse(sound.get(), sound -> {
            log.debug("playing sound: {}", sound.toString());
            mediaPlayer.setStartTime(Duration.ZERO);
            mediaPlayer.seek(Duration.ZERO);
            mediaPlayer.play();
        }, () -> log.warn("sound not set"));
    }

    public SoundResource getSound() {
        return sound.get();
    }

    public ObjectProperty<SoundResource> soundProperty() {
        return sound;
    }

    public void setSound(SoundResource sound) {
        this.sound.set(sound);
        this.mediaPlayer = Optional.ofNullable(sound)
                .map(SoundResource::get)
                .map(Object::toString)
                .map(Media::new)
                .map(MediaPlayer::new)
                .orElse(null);
    }
}

File: src/main/kotlin/io/github/dylmeadows/eontimer/service/action/SoundPlayer.kt
package io.github.dylmeadows.eontimer.service.action

import io.github.dylmeadows.eontimer.model.resource.BASE_RESOURCE_PATH
import io.github.dylmeadows.eontimer.model.resource.SoundResource
import io.github.dylmeadows.eontimer.model.settings.ActionSettingsModel
import io.github.dylmeadows.eontimer.util.asFlux
import javafx.scene.media.Media
import javafx.scene.media.MediaPlayer
import javafx.util.Duration
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.javafx.JavaFx
import kotlinx.coroutines.launch
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.net.URL
import javax.annotation.PostConstruct

@Component
class SoundPlayer @Autowired constructor(
    private val actionSettings: ActionSettingsModel) {

    private lateinit var mediaPlayer: MediaPlayer

    @PostConstruct
    private fun initialize() {
        actionSettings.soundProperty.asFlux()
            .map(SoundResource::get)
            .map(::createMediaPlayer)
            .subscribe { mediaPlayer = it }
        // NOTE: this buffers the MediaPlayer.
        // Without this buffering, the first time
        // audio is played, it is delayed.
        GlobalScope.launch(Dispatchers.JavaFx) {
            val path = "$BASE_RESOURCE_PATH/sounds/silence.wav"
            val resource = javaClass.classLoader.getResource(path)
            createMediaPlayer(resource).play()
        }
    }

    private fun createMediaPlayer(url: URL): MediaPlayer =
        MediaPlayer(Media(url.toExternalForm()))

    fun play() {
        mediaPlayer.startTime = Duration.ZERO
        mediaPlayer.seek(Duration.ZERO)
        mediaPlayer.play()
    }

    fun stop() {
        mediaPlayer.stop()
    }
}